cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java

```diff
@@ -897,22 +916,34 @@ private boolean scanApplicationsWithReorder(ScanApplicationsRequest request,
   @Override
   public Map<ApplicationId, ApplicationMeta> getApplications(Collection<ApplicationId> ids) {
     return TransactionRunners.run(transactionRunner, context -> {
-      return getAppMetadataStore(context).getApplicationsForAppIds(ids);
+      Map<ApplicationId, ApplicationMeta> appsMap = getAppMetadataStore(
+          context).getApplicationsForAppIds(ids);
+      for (Map.Entry<ApplicationId, ApplicationMeta> entry : appsMap.entrySet()) {
+        ApplicationMeta appMeta = entry.getValue();
+        SourceControlMeta sourceControlMeta = getSourceControlMetadataStore(context)
+            .getNamespaceSourceControlMeta(entry.getKey());
+        entry.setValue(new ApplicationMeta(appMeta.getId(), appMeta.getSpec(), appMeta.getChange(),
```

Comment:
In general, try to avoid mutating collections, as it makes it much more difficult to reason about the contents of the collection, and it can break if the underlying collection implementation is not modifiable.

It also wouldn't be great to create a separate `Map<ApplicationId, SourceControlMeta>`, which is probably what you're trying to avoid here. In this particular case, we could potentially pass a function into the getApplicationsForAppIds() method that returns the SourceControlMeta for a particular app id. This would have the additional benefit that it would be clear from the AppMetadataStore API that the SourceControlMeta lives in a separate table, even though it's a field in the ApplicationMeta class.

Reply:
Noted. Done